### PR TITLE
[NVIDIA] Add the sliding window to `jax.nn.dot_product_attention`

### PR DESCRIPTION
**Pull request by kaixih**
This PR enhances the `jax.nn.dot_product_attention` API by adding support for sliding window attention. A new `local_window_size` option is introduced, allowing users to define both left and right window sizes. To enable the cuDNN backend, this PR relies on the ongoing [XLA update](https://github.com/openxla/xla/pull/16358) and the corresponding [JAX cuDNN API changes](https://github.com/google/jax/pull/23203).

Additionally, this PR builds on top of [PR #23223](https://github.com/google/jax/pull/23223), which reduces the number of tests for the API.

cc. @Cjkkkk

### CODE REVIEW

**Comment by Cjkkkk** [comment]
The cuDNN API change PR: https://github.com/google/jax/pull/23284

File: jax/_src/nn/functions.py [hunk 1]
```diff
@@ -1008,9 +1032,21 @@ def _check_shape_and_dtype(t: Array | None, shape: Sequence[int],
         mask_type = MaskType.CAUSAL
       elif query_seq_lengths is not None:
         mask_type = MaskType.PADDING
+      # CuDNN supports only the left window with an exclusive boundary. So, it
+      # can be used only when the window is left-only or when causal mask is
+      # enabled.
+      sliding_window = 0
```
[the remark this answers is not in the document]
Comment (kaixih) [reply jax/_src/nn/functions.py]: Done. PTAL.

**Comment by kaixih** [comment]
Rebased to resolve conflicts.

@superbobry @sbodenstein Can you take a look?

Note, this PR depends on the this PR: https://github.com/google/jax/pull/23247#issuecomment-2313679318

**Comment by fazega** [comment]
Is the plan to include this in the next release? Useful and important feature. @superbobry

**Comment by superbobry** [comment]
Sorry for the delay, I want @sbodenstein to have a look first.

File: jax/_src/nn/functions.py [hunk 2]
```diff
@@ -1008,9 +1032,20 @@ def _check_shape_and_dtype(t: Array | None, shape: Sequence[int],
         mask_type = MaskType.CAUSAL
       elif query_seq_lengths is not None:
         mask_type = MaskType.PADDING
+      # CuDNN supports only the left window with an exclusive boundary with
+      # causal mask is enabled.
+      sliding_window = None
+      if local_window_size is not None:
+        l_window, r_window = local_window_size
+        if r_window == 0 or mask_type == MaskType.CAUSAL:
```
Comment (sbodenstein) [inline jax/_src/nn/functions.py:1040]: If r_window != 0 and causal attention is used, then the r_window seems to be silently ignored. Shouldn't this fail instead and tell the user that the combination is not allowed?

File: jax/_src/nn/functions.py [hunk 3]
```diff
@@ -1008,9 +1032,20 @@ def _check_shape_and_dtype(t: Array | None, shape: Sequence[int],
         mask_type = MaskType.CAUSAL
       elif query_seq_lengths is not None:
         mask_type = MaskType.PADDING
+      # CuDNN supports only the left window with an exclusive boundary with
+      # causal mask is enabled.
```
Comment (sbodenstein) [inline jax/_src/nn/functions.py:1036]: Nit: language doesn't make sense. Maybe 'with an exclusive boundary when causal mask is enabled'?

File: jax/_src/nn/functions.py [hunk 4]
```diff
@@ -1008,9 +1032,20 @@ def _check_shape_and_dtype(t: Array | None, shape: Sequence[int],
...
+      sliding_window = None
+      if local_window_size is not None:
+        l_window, r_window = local_window_size
+        if r_window == 0 or mask_type == MaskType.CAUSAL:
+          sliding_window = l_window + 1
+        else:
+          raise ValueError(f"cuDNN doesn't support right window: {r_window}")
```
Comment (sbodenstein) [inline jax/_src/nn/functions.py:1043]: This is not specific enough. I think cuDNN works with right window as long as causal attention is not used?

File: jax/_src/nn/functions.py (hunk 2 above)
Comment (sbodenstein) [reply jax/_src/nn/functions.py:1040]: It will also fail when r_window != 0 and non causal masking is used. The comment above suggests this should be fine. Is that allowed?

File: jax/_src/nn/functions.py [hunk 5]
```diff
@@ -785,6 +785,14 @@ def _get_causal_mask(T, S):
   mask = jnp.tril(jnp.ones((T, S), dtype=jnp.bool_))
   return mask[None, None, :, :]
 
+def _get_window_mask(T, S, local_window_size):
```
[the remark this answers is not in the document]
Comment (kaixih) [reply jax/_src/nn/functions.py]: Done.

File: jax/_src/nn/functions.py (hunk 2 above)
Comment (kaixih) [reply jax/_src/nn/functions.py:1040]: > If r_window != 0 and causal attention is used, then the r_window seems to be silently ignored. Shouldn't this fail instead and tell the user that the combination is not allowed?

No, it should work. This is intentionally, since the causal mask will effectively cover all the r_window.

> It will also fail when r_window != 0 and non causal masking is used. The comment above suggests this should be fine. Is that allowed?

Yes, this will fail. I think the comment above focuses on the behavior of cudnn, who only supports l_window + causal mask. And these branches try to translate the (l_window, r_window) + causal_mask to the cudnn with just one relaxation, i.e. when r_window is not 0 but causal is set, we are pretty sure the r_window could be safely ignored.

File: jax/_src/nn/functions.py (hunk 3 above)
Comment (kaixih) [reply jax/_src/nn/functions.py:1036]: Done.

File: jax/_src/nn/functions.py (hunk 4 above)
Comment (kaixih) [reply jax/_src/nn/functions.py]: Not really. cudnn requires causal mask and only works for l_window. Added more text in the error message.

File: jax/_src/nn/functions.py [hunk 6]
```diff
@@ -890,6 +907,7 @@ def dot_product_attention(
     is_causal: bool = False,
     query_seq_lengths: ArrayLike | None = None,
     key_value_seq_lengths: ArrayLike | None = None,
+    local_window_size: tuple[int, int] | None = None,
```
Comment (sbodenstein) [inline jax/_src/nn/functions.py:910]: What about making this either an int or a tuple of ints? If single int, then equivalent to symmetric window.

Comment (sbodenstein) [reply jax/_src/nn/functions.py:910]: Flax supports this for eg. convolution kernels.

Comment (kaixih) [reply jax/_src/nn/functions.py]: Sure. Done.

**Comment by kaixih** [comment]
Force push to resolve conflicts.

**Comment by sbodenstein** [comment]
LGTM

**Review by superbobry — APPROVED** [review]

**Comment by kaixih** [comment]
Gentle ping. Seems the merging is blocked. @superbobry
